### PR TITLE
Add CollectionSerializer and ready serializer API for use in configuration.

### PR DESCRIPTION
**Pull request by orbyfied**
- I had to write a custom `CollectionSerializer` as I couldn't find a brigadier one, and it is much easier to maintain this way.

### CODE REVIEW

File: src/main/java/com/domain/redstonetools/features/commands/QuickTpFeature.java [hunk 1]
```diff
@@ -29,6 +30,7 @@ protected int execute(ServerCommandSource source) throws CommandSyntaxException
         var player = source.getPlayer();
 
         var targetPosition = getTargetPosition(player);
+        player.onTeleportationDone();
```
Comment (Matthias1590) [inline src/main/java/com/domain/redstonetools/features/commands/QuickTpFeature.java:33]: Just curious, what does this method do? Also shouldnt it be called after the teleportation is done as opposed to before the teleportation occurs?

File: src/main/java/com/domain/redstonetools/features/arguments/StringSerializer.java [hunk 2]
```diff
@@ -1,31 +1,33 @@
...
 
-    public static StringSerializer string() {
-        return INSTANCE_STRING;
+    public static StringSerializer greedy() {
+        return INSTANCE_GREEDY;
     }
 
-    public static StringSerializer greedyString() {
-        return INSTANCE_GREEDY_STRING;
+    public StringSerializer(ArgumentType<String> argumentType) {
```
Comment (Matthias1590) [inline src/main/java/com/domain/redstonetools/features/arguments/StringSerializer.java:24]: Why make the constructor public? Don't we want just one instance?

File: src/main/java/com/domain/redstonetools/features/arguments/BrigadierSerializer.java [hunk 3]
```diff
@@ -10,26 +10,29 @@
 import java.util.Collection;
 import java.util.concurrent.CompletableFuture;
 
-public abstract class BrigadierSerializer<T> extends TypeSerializer<T> {
-    private final ArgumentType<T> argType;
+public abstract class BrigadierSerializer<T, S> extends TypeSerializer<T, S> {
 
-    protected BrigadierSerializer(Class<T> clazz, ArgumentType<T> argType) {
+    // the wrapped brigadier argument type
+    final ArgumentType<T> argumentType;
```
Comment (Matthias1590) [inline src/main/java/com/domain/redstonetools/features/arguments/BrigadierSerializer.java:16]: Why is this not private?

File: src/main/java/com/domain/redstonetools/features/arguments/BoolSerializer.java [hunk 4]
```diff
@@ -1,20 +1,23 @@
...
-    private BoolSerializer() {
-        super(Boolean.class, BoolArgumentType.bool());
-    }
+    static final BoolSerializer INSTANCE = new BoolSerializer(BoolArgumentType.bool());
 
     public static BoolSerializer bool() {
         return INSTANCE;
     }
 
+    public BoolSerializer(ArgumentType<Boolean> argumentType) {
```
Comment (Matthias1590) [inline src/main/java/com/domain/redstonetools/features/arguments/BoolSerializer.java:14]: This shouldnt be public

File: src/main/java/com/domain/redstonetools/features/arguments/BoolSerializer.java [hunk 5]
```diff
@@ -1,20 +1,23 @@
...
 import com.mojang.brigadier.arguments.BoolArgumentType;
 
-public class BoolSerializer extends BrigadierSerializer<Boolean> {
-    private static final BoolSerializer INSTANCE = new BoolSerializer();
+public class BoolSerializer extends StringBrigadierSerializer<Boolean> {
 
-    private BoolSerializer() {
-        super(Boolean.class, BoolArgumentType.bool());
-    }
+    static final BoolSerializer INSTANCE = new BoolSerializer(BoolArgumentType.bool());
```
Comment (Matthias1590) [inline src/main/java/com/domain/redstonetools/features/arguments/BoolSerializer.java:8]: This can just be private

**Review by Matthias1590 — CHANGES_REQUESTED** [review]

File: src/main/java/com/domain/redstonetools/features/arguments/DoubleSerializer.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
+package com.domain.redstonetools.features.arguments;
+
+import com.mojang.brigadier.arguments.ArgumentType;
+import com.mojang.brigadier.arguments.DoubleArgumentType;
+
+public class DoubleSerializer extends StringBrigadierSerializer<Double> {
```
[the remark this answers is not in the document]
Comment (orbyfied) [reply src/main/java/com/domain/redstonetools/features/arguments/DoubleSerializer.java:6]: If we need it in the future we have it, there's no reason to remove it really.

[the remark this answers is not in the document]
Comment (orbyfied) [reply src/main/java/com/domain/redstonetools/features/arguments/DoubleSerializer.java:6]: It's one of the primitives

File: src/main/java/com/domain/redstonetools/features/arguments/BoolSerializer.java [hunk 7]
```diff
@@ -5,13 +5,13 @@
...
 
-    static final BoolSerializer INSTANCE = new BoolSerializer(BoolArgumentType.bool());
+    private static final BoolSerializer INSTANCE = new BoolSerializer(BoolArgumentType.bool());
 
     public static BoolSerializer bool() {
         return INSTANCE;
     }
 
-    public BoolSerializer(ArgumentType<Boolean> argumentType) {
+    BoolSerializer(ArgumentType<Boolean> argumentType) {
```
Comment (Matthias1590) [inline src/main/java/com/domain/redstonetools/features/arguments/BoolSerializer.java:14]: Should be private, not package private

File: src/main/java/com/domain/redstonetools/features/arguments/DoubleSerializer.java [hunk 8]
```diff
@@ -19,7 +19,7 @@ public static DoubleSerializer doubleArg(double min, double max) {
         return new DoubleSerializer(DoubleArgumentType.doubleArg(min, max));
     }
 
-    public DoubleSerializer(ArgumentType<Double> argumentType) {
+    DoubleSerializer(ArgumentType<Double> argumentType) {
```
Comment (Matthias1590) [inline src/main/java/com/domain/redstonetools/features/arguments/DoubleSerializer.java:22]: 📦 🔒 ->🔒

File: src/main/java/com/domain/redstonetools/features/arguments/FloatSerializer.java [hunk 9]
```diff
@@ -19,7 +19,7 @@ public static FloatSerializer floatArg(float min, float max) {
         return new FloatSerializer(FloatArgumentType.floatArg(min, max));
     }
 
-    public FloatSerializer(ArgumentType<Float> argumentType) {
+    FloatSerializer(ArgumentType<Float> argumentType) {
```
Comment (Matthias1590) [inline src/main/java/com/domain/redstonetools/features/arguments/FloatSerializer.java:22]: private

File: src/main/java/com/domain/redstonetools/features/arguments/IntegerSerializer.java [hunk 10]
```diff
@@ -19,7 +19,7 @@ public static IntegerSerializer integer(int min, int max) {
         return new IntegerSerializer(IntegerArgumentType.integer(min, max));
     }
 
-    public IntegerSerializer(ArgumentType<Integer> argumentType) {
+    IntegerSerializer(ArgumentType<Integer> argumentType) {
```
Comment (Matthias1590) [inline src/main/java/com/domain/redstonetools/features/arguments/IntegerSerializer.java:22]: private

File: src/main/java/com/domain/redstonetools/features/arguments/LongSerializer.java [hunk 11]
```diff
@@ -19,7 +19,7 @@ public static LongSerializer longArg(long min, long max) {
         return new LongSerializer(LongArgumentType.longArg(min, max));
     }
 
-    public LongSerializer(ArgumentType<Long> argumentType) {
+    LongSerializer(ArgumentType<Long> argumentType) {
```
Comment (Matthias1590) [inline src/main/java/com/domain/redstonetools/features/arguments/LongSerializer.java:22]: private

File: src/main/java/com/domain/redstonetools/features/arguments/StringSerializer.java [hunk 12]
```diff
@@ -17,11 +17,11 @@ public static StringSerializer word() {
...
     }
 
-    public static StringSerializer greedy() {
-        return INSTANCE_GREEDY;
+    public static StringSerializer greedyString() {
+        return INSTANCE_GREEDY_STRING;
     }
 
-    public StringSerializer(ArgumentType<String> argumentType) {
+    StringSerializer(ArgumentType<String> argumentType) {
```
Comment (Matthias1590) [inline src/main/java/com/domain/redstonetools/features/arguments/StringSerializer.java:24]: private

**Review by Matthias1590 — CHANGES_REQUESTED** [review]

**Comment by orbyfied** [comment]
@Matthias1590 Changed implemented

**Pull request merged**